src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,286 @@
+#include "mfa_hierarchy.h"
+
+#include <google/protobuf/stubs/casts.h>
+
+#include <Eigen/Dense>
+#include <cassert>
+#include <iostream>
+#include <stan/math/prim/prob.hpp>
+#include <vector>
+
+#include "algorithm_state.pb.h"
+#include "hierarchy_prior.pb.h"
+#include "ls_state.pb.h"
+#include "src/utils/proto_utils.h"
+#include "src/utils/rng.h"
+
+double MFAHierarchy::like_lpdf(const Eigen::RowVectorXd& datum) const {
+  return stan::math::multi_normal_lpdf(
+      datum, state.mu,
+      state.lambda * state.lambda.transpose() +
+          Eigen::MatrixXd(state.psi.asDiagonal()));
+}
+
+MFA::State MFAHierarchy::draw(const MFA::Hyperparams& params) {
+  auto& rng = bayesmix::Rng::Instance().get();
+  MFA::State out;
+  out.mu = params.mutilde;
+  out.psi = params.beta / (params.alpha0 + 1.);
+  out.eta = Eigen::MatrixXd::Zero(card, params.q);
+  out.lambda = Eigen::MatrixXd::Zero(dim, params.q);
+
+  for (size_t j = 0; j < dim; j++) {
+    out.mu[j] =
+        stan::math::normal_rng(params.mutilde[j], sqrt(params.phi), rng);
+
+    out.psi[j] = stan::math::inv_gamma_rng(params.alpha0, params.beta[j], rng);
+
+    for (size_t i = 0; i < params.q; i++) {
+      out.lambda(j, i) = stan::math::normal_rng(0, 1, rng);
+    }
+  }
+
+  for (size_t i = 0; i < card; i++) {
+    for (size_t j = 0; j < params.q; j++) {
+      out.eta(i, j) = stan::math::normal_rng(0, 1, rng);
+    }
+  }
+
+  out.psi_inverse = out.psi.cwiseInverse().asDiagonal();
+
+  return out;
+}
+
+void MFAHierarchy::initialize_state() {
+  std::cout << "init state" << std::endl;
+  state.mu = hypers->mutilde;
+  state.psi = hypers->beta / (hypers->alpha0 + 1.);
+  state.eta = Eigen::MatrixXd::Zero(card, hypers->q);
+  state.lambda = Eigen::MatrixXd::Zero(dim, hypers->q);
+  state.psi_inverse = state.psi.cwiseInverse().asDiagonal();
+}
+
+void MFAHierarchy::initialize_hypers() {
+  if (prior->has_fixed_values()) {
+    // Set values
+    hypers->mutilde = bayesmix::to_eigen(prior->fixed_values().mutilde());
+    dim = hypers->mutilde.size();
+    hypers->beta = bayesmix::to_eigen(prior->fixed_values().beta());
+    hypers->phi = prior->fixed_values().phi();
+    hypers->alpha0 = prior->fixed_values().alpha0();
+    hypers->q = prior->fixed_values().q();
+
+    // Automatic initialization
```

Comment:
Now, since we have the whole dataset inside the hierarchy, we can initialize the hyperparameters with respect to the sample mean and sample covariance matrix.